### DISCUSSION TITLE
next steps for `MimeBundleDescriptor`: custom messages / non-synchronized state

### DISCUSSION THREAD

**Discussion by manzt** [manzt/anywidget#294] [Ideas]
The `MimeBundeDescriptor` introduces a more library agnostic connection between python model and JS view, but currently doesn't support two important features from `ipywidgets.Widget`. First a mechanism to both send _and_ receive messages from the front end. The latter could be simple enough to implement with a special method (akin to `_anywidget_get_state`):

```python
class Foo:
  _repr_mimebundle_ = MimeBundleDescriptor()
  def _anywidget_handle_custom_msg(self, msg):
      ...
```

But I'm having trouble thinking of a way to allow for custom messages to be sent. Right now this can be accomplished with:

```python
class Foo:
  _repr_mimebundle_ = MimeBundleDescriptor()

Foo()._repr_mimebundle_._comm.send(...) 
# probably do not want to expose Comm directly bc custom messages have a particular format
```

but I'd be ok with adding a `send` method to `ReprMimeBundle` directly (like `send_state`):

```python
class Foo:
  _repr_mimebundle_ = MimeBundleDescriptor()

Foo()._repr_mimebundle_.send(contents, buffers
# alias for ReprMimeBundle._comm.send({"method": "custom", "content": content}, buffers=buffers)
```

Of course, using the `_repr_mimbundle_` name is a bit verbose, but I like how it makes it clear how these messages are only received if the front end is activated.

Second, we should have a way of marking attributes which shouldn't automatically be synchronized with the frontend. This would be the opposite of `sync=True` for traitlets where by default everything is synchronized.

**Comment by tlambert03** [comment]
I'm not sure I see the `_repr_mimebundle_ = MimeBundleDescriptor()` as a public facing thing.  It _could_ be, but it could also be the building blocks for a thin base class.

For example, you could make `AnyWidget` itself inherit from HasTraits instead of DOMWidget (or dataclass, or msgspec, etc...), and then the user doesn't particularly care about the details of the descriptor.  Something like:

```py
class AnyWidget(t.HasTraits):
    _repr_mimebundle_ = MimeBundleDescriptor()
    def on_msg(self, msg):
        return NotImplemented
    def send(self, contents, buffers):
        ...
```

also, note that you can always add an alias for `_repr_mimebundle_` if you want:

```python
class AnyWidget:
    _repr_mimebundle_ = MimeBundleDescriptor()
    alias = _repr_mimebundle_
```

`_repr_mimebundle_` is, of course, the *bare minimum* name required to implement the Jupyter API, but you can add more friendly names to access the same object if you want

**Comment by manzt** [comment]
Might also be worth exploring the [`comm`](https://pypi.org/project/comm/) package instead of relying on ipykernel? https://github.com/jupyter-widgets/ipywidgets/pull/3533 (not sure of the motivation behind this change).

**Comment by manzt** [comment]
Switched over to `comm` in #119 , hoping to expand on this effort soon. Would be really neat not require ipywidgets as a dep.

**Comment by manzt** [comment]
> For example, you could make AnyWidget itself inherit from HasTraits instead of DOMWidget 

@tlambert03, dang I'm seeing the motivation for this only now. You had such foresight! ipywidgets is a huge dependency, and I'd like to experiment with a pure `traitlets` + our descriptor for the default implementation. Maybe we could have an import from:

```py
from anywidget.ipywidgets import AnyWidget
```

and move ipywidgets towards an optional dependency, and then only expose a few APIs like `on_msg`, `send` and some helpers for exporting state as dicts.

I've been preferring the traitlets API since the Python side is consistent with connecting to existing ipywidgets, although I much prefer the `psygnal` API. I'm realizing similar to the JS side, anywidget has the opportunity to push widgets forward (and since not that many people are experienced with widgets, familiarity with existing APIs isn't as big of a concern as I'd anticipated), although I want to avoid creating too much friction with the existing ecosystem (i.e., anywidgets play nice with other traditional jupyter widgets).
